packages/IntegrationTest/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -0,0 +1,195 @@
+
+  m_commandListener->OnTestCommand([weakThis{get_weak()}](
+      const TestCommand &command, TestCommandResponse response) noexcept {
+    if (auto strongThis = weakThis.get()) {
+      strongThis->OnTestCommand(command, std::move(response));
+    }
+  });
+}
+
+void TestHostHarness::SetRootView(ReactRootView &&rootView) noexcept {
+  m_rootView = std::move(rootView);
+}
+
+void TestHostHarness::SetReactHost(ReactNativeHost &&reactHost) noexcept {
+  reactHost.InstanceSettings().RedBoxHandler(m_redboxHandler);
+  m_instanceLoadedRevoker = reactHost.InstanceSettings().InstanceLoaded(
+      winrt::auto_revoke, [weakThis{get_weak()}](const auto & /*sender*/, const auto &args) noexcept {
+        auto strongThis = weakThis.get();
+        if (!strongThis) {
+          return;
+        }
+
+        ReactContext context(args.Context());
+
+        strongThis->m_context = context;
+        strongThis->StartListening();
+
+        context.Notifications().Subscribe(
+            TestModule::TestCompletedEvent(),
+            context.UIDispatcher(),
+            [weakThis](const auto & /*sender*/, ReactNotificationArgs<void> /*args*/) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestCompleted();
+              }
+            });
+
+        context.Notifications().Subscribe(
+            TestModule::TestPassedEvent(),
+            context.UIDispatcher(),
+            [weakThis](const auto & /*sender*/, ReactNotificationArgs<bool> args) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestPassed(*args.Data());
+              }
+            });
+      });
+}
+
+winrt::fire_and_forget TestHostHarness::StartListening() noexcept {
+  auto listenResult = co_await m_commandListener->StartListening();
+
+  switch (listenResult) {
+    case ListenResult::Success:
+      co_return;
+
+    case ListenResult::AddressInUse:
+      ShowJSError("Test harness address is already in use. Is there another instance running?");
+      co_return;
+
+    case ListenResult::OtherError:
+      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
+      co_return;
+
+    default:
+      ShowJSError("Unexpected ListenResult from TestCommandListener.");
+      co_return;
+  }
+}
+
+winrt::fire_and_forget TestHostHarness::OnTestCommand(TestCommand command, TestCommandResponse response) noexcept {
+  // Keep ourselves alive while we have a test command
+  auto strongThis = get_strong();
+
+  switch (command.Id) {
+    case TestCommandId::RunTestComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+
+      // We might see multiple JS events for exceptions, failing tests, etc.
+      // Flush everything pending from the JS thread to UI thread before
+      // continuing to ensure all events from the past command are handled.
+      co_await FlushJSQueue();
```

Comment:
This still allows some bleed-through, since more JS work corresponding to the previous component can be scheduled after this. This mostly affects the RedBoxHandler, since we need to discern between an exception that means a test doesn't pass vs exceptions happening while the previous test is still doing their thing. Exception bleedthrough isn't the worst bug though. Ideally we could tie into componentWIllUnmount as a safe point instead of idle but I'm not sure if that's possible.

Reply:
Looking at the mechanics of this a bit more, from what I can tell, changing the root view component name puts a message in a serial queue to then tell the instance to uninit the root view, which will eventually call into JS to do the uninit and then wait until the JS thread processes it.

The non-ABI ReactViewHost classes offer futures for this. Maybe I should be thinking about how to add lifecycle callbacks to the root view.